NetCasbin/Enforcer.cs

```diff
@@ -21,629 +30,647 @@ public Enforcer(string modelPath, string policyPath)
         {
         }
 
-        public Enforcer(string modelPath, IAdapter adapter)
-            : this(NewModel(modelPath, string.Empty), adapter)
+        public Enforcer(string modelPath, IAdapter adapter = null)
+            : this(Casbin.Model.Model.CreateDefaultFromFile(modelPath), adapter)
         {
-            this.modelPath = modelPath;
+            ModelPath = modelPath;
         }
 
-        public Enforcer(Model.Model model, IAdapter adapter)
+        public Enforcer(Model.Model model, IAdapter adapter = null)
         {
-            this.adapter = adapter;
-            watcher = null;
+            if (adapter is not null)
+            {
+                SetAdapter(adapter);
+            }
+
             SetModel(model);
-            Initialize();
             LoadPolicy();
         }
 
-        public Enforcer(Model.Model m)
-            : this(m, null)
-        {
-        }
+        #region Options
+        public bool Enabled { get; private set; } = true;
+        public bool AutoSave { get; private set; } = true;
+        public bool AutoBuildRoleLinks { get; private set; } = true;
+        public bool AutoNotifyWatcher { get; private set; } = true;
+        #endregion
 
-        public Enforcer(string modelPath)
-            : this(modelPath, string.Empty)
-        {
-        }
+        #region Extensions
+        public IEffector Effector { get; private set; } = new DefaultEffector();
+        public Model.Model Model { get; private set; }
+        public IAdapter Adapter { get; private set; }
+        public IWatcher Watcher { get; private set; }
+        public IRoleManager RoleManager { get; private set; } = new DefaultRoleManager(10);
+#if !NET45
+        public ILogger Logger { get; set; }
+#endif
+        #endregion
 
-        [Obsolete("The method will be removed at next mainline version, you can see https://github.com/casbin/Casbin.NET/issues/116 to know more information.")]
-        public Enforcer(string modelPath, string policyFile, bool enableLog)
-            : this(modelPath, new FileAdapter(policyFile))
-        {
+        public string ModelPath { get; private set; }
+        public bool IsFiltered => Adapter is IFilteredAdapter {IsFiltered: true};
 
-        }
+        public IExpressionHandler ExpressionHandler { get; private set; }
```

Comment:
Because `internal` is not supported in the interface before netstatandard2.1 (default interface implement), but it needs to be called at the `AddFunction` extension method.
And I also think this property should allow users to view current params, clear cache, or set their own implementation to enhance scalability.
Sure, we also can add the `AddFunction` method to `IEnforcer` to keep this property internal.

Reply:
@xcaptain